application/src/main/java/org/togetherjava/tjbot/commands/moderation/report/ReportCommand.java

```diff
@@ -0,0 +1,175 @@
+ * {@link Config#getModMailChannelPattern()} ()}.
+ */
+
+public final class ReportCommand extends BotCommandAdapter implements MessageContextCommand {
+
+    private static final Logger logger = LoggerFactory.getLogger(ReportCommand.class);
+    private static final String COMMAND_NAME = "report";
+    private static final String MESSAGE_INPUT = "message";
+    private static final int COOLDOWN_DURATION_VALUE = 3;
+    private static final ChronoUnit COOLDOWN_DURATION_UNIT = ChronoUnit.MINUTES;
+    private static final Color AMBIENT_COLOR = Color.BLACK;
+    private final Cache<Long, Instant> authorToLastReportInvocation = createCooldownCache();
+    private final Predicate<String> modMailChannelNamePredicate;
+    private final String configModMailChannelPattern;
+    private String reportedMessage;
+    private String reportedMessageUrl;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param config to get the channel to forward reports to.
+     */
+    public ReportCommand(Config config) {
+        super(Commands.message(COMMAND_NAME), CommandVisibility.GUILD);
+
+        modMailChannelNamePredicate =
+                Pattern.compile(config.getModMailChannelPattern()).asMatchPredicate();
+
+        configModMailChannelPattern = config.getModMailChannelPattern();
+    }
+
+    private Cache<Long, Instant> createCooldownCache() {
+        return Caffeine.newBuilder()
+            .maximumSize(1_000)
+            .expireAfterAccess(COOLDOWN_DURATION_VALUE, TimeUnit.of(COOLDOWN_DURATION_UNIT))
+            .build();
+    }
+
+    @Override
+    public void onMessageContext(MessageContextInteractionEvent event) {
+        reportedMessage = event.getTarget().getContentRaw();
+        reportedMessageUrl = event.getTarget().getJumpUrl();
+
+        TextInput body = TextInput.create(MESSAGE_INPUT, "Message", TextInputStyle.PARAGRAPH)
+            .setPlaceholder("Reason for reporting")
+            .setRequiredRange(10, 200)
+            .build();
+
+        Modal modal = Modal.create(generateComponentId(), "Report").addActionRow(body).build();
+
+        event.replyModal(modal).queue();
+    }
+
+    @Override
+    public void onModalSubmitted(ModalInteractionEvent event, List<String> args) {
+        long userID = event.getUser().getIdLong();
+
+        if (handleIsOnCooldown(userID, event)) {
+            return;
+        }
+        authorToLastReportInvocation.put(userID, Instant.now());
+        event.deferReply().setEphemeral(true).queue();
+
+        long guildID = Objects.requireNonNull(event.getGuild(), "Could not retrieve the guildId.")
+            .getIdLong();
+        Optional<TextChannel> modMailAuditLog = getModMailChannel(event.getJDA(), guildID);
+        if (modMailAuditLog.isEmpty()) {
+            logger.warn(
+                    "Cannot find the designated modmail channel in server by id {} with the pattern {}",
+                    guildID, configModMailChannelPattern);
+            return;
+        }
```

Comment:
might want to add a `handleRequireModMailChannel` helper method to simplify this method and reduce its scope.

Reply:
I kind of tried adding the whole thing into the helper method, but extracting it , just goes that I add a method within the modmailAuditLog = handleRequire... . 

When I wrap the whole thing, other methods still need the modmailAuditLog passed to them

So I am not sure now to do this?

Reply:
at which point u can also just transform the `getModMailChannel` into the `handleRequireModMailChannel` method instead. i.e. give it also the responsibility to do the log.

Reply:
It was unclear because I still had to use on call site handleRequest... and within the method getModmail... 

Thought that it was not suppose to be done that way. But I am glad its cleared out. Thank you

Reply:
Did you mean by transforming just to do 

from this 

`getModMailChannel(jda, guildID)`

to this
```java
Optional<TextChannel> modMailAuditLog = jda.getGuildById(guildID)
                                                   .getTextChannelCache()
                                                   .stream()
                                                   .filter(channel -> modMailChannelNamePredicate.test(channel.getName()))
                                                   .findAny();
```